FirebaseSessions/Sources/SettingsDownloadClient.swift

```diff
@@ -0,0 +1,99 @@
+// limitations under the License.
+
+import Foundation
+
+@_implementationOnly import GoogleUtilities
+
+protocol SettingsDownloadClient {
+  func fetch(completion: @escaping (Result<[String: Any], Error>) -> Void)
+}
+
+class SettingsDownloader: SettingsDownloadClient {
+  private let appInfo: ApplicationInfoProtocol
+  private let identifiers: IdentifierProvider
+  private let installations: InstallationsProtocol
+
+  init(appInfo: ApplicationInfoProtocol, identifiers: IdentifierProvider,
+       installations: InstallationsProtocol) {
+    self.appInfo = appInfo
+    self.identifiers = identifiers
+    self.installations = installations
+  }
+
+  func fetch(completion: @escaping (Result<[String: Any], Error>) -> Void) {
+    guard let url = url else {
+      completion(.failure(FirebaseSessionsError.SettingsError("Invalid URL")))
+      return
+    }
+
+    buildRequest(url: url) { result in
+      switch result {
+      case let .success(request):
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+          if let data = data {
+            if let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
+              completion(.success(dict))
+            } else {
+              completion(.failure(FirebaseSessionsError
+                  .SettingsError("Failed to parse JSON to dictionary")))
+            }
+          } else if let error = error {
+            completion(.failure(FirebaseSessionsError.SettingsError(error.localizedDescription)))
+          }
+        }
+        task.resume()
+      case let .failure(error):
+        completion(.failure(FirebaseSessionsError.SettingsError(error.localizedDescription)))
+      }
+    }
+  }
+
+  private var url: URL? {
+    var components = URLComponents()
+    components.scheme = "https"
+    components.host = "firebase-settings.crashlytics.com"
+    components.path = "/spi/v2/platforms/android/gmp/\(appInfo.appID)/settings"
+    components.queryItems = [
+      URLQueryItem(name: "build_version", value: appInfo.appBuildVersion),
+      URLQueryItem(name: "display_version", value: appInfo.appDisplayVersion),
+    ]
+    return components.url
+  }
+
+  private func buildRequest(url: URL, completion: @escaping (Result<URLRequest, Error>) -> Void) {
+    installations.installationID { [self] result in
+      switch result {
+      case let .success(fiid):
+        var request = URLRequest(url: url)
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+        request.setValue(fiid, forHTTPHeaderField: "X-Crashlytics-Installation-ID")
+        request.setValue(appInfo.deviceModel, forHTTPHeaderField: "X-Crashlytics-Device-Model")
+        request.setValue(
+          FIRSESGetSysctlEntry("kern.osversion"),
+          forHTTPHeaderField: "X-Crashlytics-OS-Build-Version"
+        )
+        request.setValue(
+          GULAppEnvironmentUtil.systemVersion(),
```

Comment:
To make these available in case they are useful to other parts of our code, I would recommend putting `GULAppEnvironmentUtil.systemVersion()` and `FIRSESGetSysctlEntry("kern.osversion")` into ApplicationInfo.

Reply:
Done